h2o-algos/src/test/java/hex/grid/GridSearchWithFilteringTest.java

```diff
@@ -0,0 +1,137 @@
+package hex.grid;
+
+import hex.grid.filter.KeepOnlyFirstMatchFilterFunction;
+import hex.grid.filter.PermutationFilterFunction;
+import hex.tree.gbm.GBMModel;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import water.TestUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Note: exact type of {@code Model.Parameters} is transparent to filtering functionality 
+ * so we can test on any of the existing ones.
+ */
+public class GridSearchWithFilteringTest extends TestUtil {
+
+  @BeforeClass
+  public static void setup() {
+    stall_till_cloudsize(1);
+  }
+  
+  @Test
+  public void randomGridSearchWithFunctionsToFilterOutUnnecessaryGridItems_GBMParametersGrid() {
+    Map<String, Object[]> hpGrid = new HashMap<>();
+    hpGrid.put("_ntrees", new Integer[]{10000});
+    hpGrid.put("_max_depth", new Integer[]{3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17}); 
+    hpGrid.put("_min_rows", new Integer[]{1, 5, 10, 15, 30, 100}); 
+    hpGrid.put("_learn_rate", new Double[]{0.001, 0.005, 0.008, 0.01, 0.05, 0.08, 0.1, 0.5, 0.8}); 
+    hpGrid.put("_sample_rate", new Double[]{0.50, 0.60, 0.70, 0.80, 0.90, 1.00}); 
+    hpGrid.put("_col_sample_rate", new Double[]{ 0.4, 0.7, 1.0}); 
+    hpGrid.put("_col_sample_rate_per_tree", new Double[]{ 0.4, 0.7, 1.0}); 
+    hpGrid.put("_min_split_improvement", new Double[]{1e-4, 1e-5});
+
+    int totalNumberOfPermutationsInHPGrid = 87480;
+
+    GBMModel.GBMParameters gbmParameters = new GBMModel.GBMParameters();
+
+    GridSearch.SimpleParametersBuilderFactory simpleParametersBuilderFactory = new GridSearch.SimpleParametersBuilderFactory();
+
+    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria hyperSpaceSearchCriteria = new HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria();
+
+    PermutationFilterFunction strictFilterFunction = gridItem -> {
+      return !((int) gridItem.get("_max_depth") >= 15.0 && (int) gridItem.get("_min_rows") >= 50);
+    };
+    int expectedNumberOfFilteredOutPermutations = 2916;
+
+    ArrayList<PermutationFilterFunction> filterFunctions = new ArrayList<>();
+    filterFunctions.add(strictFilterFunction);
```

Comment:
Collections.singletonList or List.of() are nicer options

Reply:
@honzasterba could you please elaborate on why these options are nicer? I would say that we can avoid random access features but almost all implementations of `List` interface provide access by index. In our case `Collection` would be enough I think as it provides `stream()` functionality.

Collections.singletonList can not be used in the following use case as we need more than one entry:
```
ArrayList<PermutationFilterFunction> filterFunctions = new ArrayList<>();
    filterFunctions.add(filterFunction1);
    filterFunctions.add(filterFunction2);
```

I don't think that carefully choosing type of List implementation here will give us a lot of benefits. Normal use case is to just have couple of entries.

Reply:
because you are creating a list with one item in it, with static import the line can be just

     List<PermutationFilterFunction> filterFunctions = singletonList(strictFilterFunction);

which is more readable IMO

Reply:
`Arrays.asList(...)` should make everyone happy.